refactor: move init_default_autocmds to lua

glacambre opened the pull request:
The original motivation for this change came from developping
https://github.com/neovim/neovim/pull/22159, which will require adding
more autocommand creation to Neovim's startup sequence.

glacambre committed:
refactor: move init_default_autocmds to lua

The original motivation for this change came from developping
https://github.com/neovim/neovim/pull/22159, which will require adding
more autocommand creation to Neovim's startup sequence.

This change requires lightly editing a test that expected no autocommand
to have been created from lua.

bfredl commented on src/nvim/main.c:
```diff
@@ -415,7 +415,13 @@ int main(int argc, char **argv)
 
-  init_default_autocmds();
+  err = ERROR_INIT;
+  o = NLUA_EXEC_STATIC("return vim._init_default_autocmds()",
```
I'd suggest merge this with the preceding `NLUA_EXEC_STATIC` call to (1) avoid the repetition of boilerplate (2) entering/exiting the luajit VM is much more expensive than having a lua function just call another lua function.

glacambre replied:
Thanks for the review. I was half-hoping that the diffs I was seeing were coming from a broken setup on my end, but it looks like they're real. I'll turn this PR into a draft, fix it, and then ping you again :)

glacambre committed:
fix(helpers): restore channel id after a call to WITH_SCRIPT_CONTEXT

In https://github.com/neovim/neovim/pull/22214, init_default_autocmds
has been turned into a lua function call to nvim_create_augroup and
nvim_create_autocmd.

This introduced a strange regression: a test in vim_spec.lua started
failing with its last_set_chan value switching from 0 to
-9223372036854775808.

It turns out that -9223372036854775808 is the value of LUA_INTERNAL_CALL
and would be inherited as last_set_chan by options set from the command
line due to the WITH_SCRIPT_CONTEXT macro not restoring the channel id
(WITH_SCRIPT_CONTEXT is used by nvim_create_augroup).

justinmk commented on test/functional/api/autocmd_spec.lua:
```diff
@@ -209,7 +209,7 @@ describe('autocmd api', function()
       local aus = meths.get_autocmds({ event = 'User', pattern = 'Test' })
       local first = aus[1]
-      eq(first.id, 1)
+      eq(true, first.id > 0)
```
just for future reference: can use ok() to get a nicer failure message.
```suggestion
      ok(first.id > 0, 'first.id > 0', first.id)
```

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -894,6 +894,26 @@ function vim._init_default_mappings()
+    pattern = 'term://*',
+    group = nvim_terminal_augroup,
+    nested = true,
+    command = "if !exists('b:term_title')|call termopen(matchstr(expand(\"<amatch>\"), '\\c\\mterm://\\%(.\\{-}//\\%(\\d\\+:\\)\\?\\)\\?\\zs.*'), {'cwd': expand(get(matchlist(expand(\"<amatch>\"), '\\c\\mterm://\\(.\\{-}\\)//'), 1, ''))})",
```
could avoid some escaping by using `[[ ... ]]`

justinmk commented:
Nice change. Merged to unblock #22159

glacambre commented:
Thanks a lot for the review!
